data/index.js

```diff
@@ -139,7 +155,23 @@ export function select( reducerKey ) {
 		return select( reducerKey )[ selectorKey ]( ...args );
 	}
 
-	return selectors[ reducerKey ];
+	const createResolver = ( selector, key ) => ( ...args ) => {
+		const result = selector( ...args );
+		const resolver = resolvers[ reducerKey ] && resolvers[ reducerKey ][ key ];
+		if ( ! resolver ) {
+			return result;
+		}
+		const isFulfilled = resolver.isFulfilled( stores[ reducerKey ].getState(), ...args );
+		if ( isFulfilled ) {
+			// TODO: clear memize cache for these sepecific args
```

Comment:
I don't think we need to do anything special here. If `fulfill` returns a promise, we can just `then` to it because it works, whether or not the request has already been completed or not.

See example: http://jsbin.com/qetusew/edit?html,js,console

Reply:
I guess it becomes a question of whether, if we want separate `isFulfilled` and `fulfill`, whether we need to keep around the cached value for `fulfill`, since'd expect future values of `isFulfilled` to return true.

Where this could be useful is if we established a convention of the selector or state returning a value indicating that it is not fulfilled: `null` or, more likely `undefined`, since we may want to express meaningful emptiness via `null`.

Reply:
The idea is that we don't want to call `fulfill` if it's already in-flight or fulfilled.

1- If it's already fulfilled, we have a dedicated API for this `isFulfilled`
2- If it's already in flights, we need to cache a value saying that `fulfilled` has already been called, we can do so by storing a map `[ args ] => promise` or we can just rely on a memoization library doing exactly that, avoid calling a function again if the args are the same.
I preferred the second option using `memize` because it seemed simpler and `memize` is already well optimized.

Now, when the selector is fulfilled, we may want to clear the memoized value (or the cached promise) because it's not necessary anymore. That's what my `TODO` comment is about.

Reply:
But I agree that returning a promise from `fulfill` is a good convention to have even if we don't use right now, it could be useful later for other purposes.